File: frontend/lib/api/approvals.ts
import { request } from "@/lib/api/client";

export type Approval = {
  id: string; run_id: string; agent_name: string; capability: string; tool_name: string;
  resource: string | null; argument_summary: Record<string, string | number | boolean | null>;
  reason: string; status: "pending" | "approved" | "rejected" | "expired" | "cancelled";
  created_at: string; expires_at: string | null;
};

export const approvalsApi = {
  list: (runId: string) => request<Approval[]>(`/runs/${runId}/approvals`),
  approve: (id: string) => request<Approval>(`/approvals/${id}/approve`, { method: "POST" }),
  reject: (id: string) => request<Approval>(`/approvals/${id}/reject`, { method: "POST" }),
};

File: app/api/schemas/approvals.py
"""Safe HTTP representations for human approval gates."""

from datetime import datetime
from pydantic import BaseModel

from app.security.approvals import ApprovalStatus
from app.security.contracts import Capability


class ApprovalResponse(BaseModel):
    id: str
    run_id: str
    parent_run_id: str | None
    agent_name: str
    capability: Capability
    tool_name: str
    resource: str | None
    argument_summary: dict[str, str | int | bool | None]
    reason: str
    status: ApprovalStatus
    created_at: datetime
    resolved_at: datetime | None
    expires_at: datetime | None
    policy_id: str
    action_fingerprint: str
